### DISCUSSION TITLE
type checking seems broken when changing a function between explicit types and inferred

### DISCUSSION THREAD

**Discussion by ehaynes99** [vercel/turborepo#836] [Help] [answered]
I'm completely new to turborepo, just trying it out for the first time. I created a new project as instructed in the README. However, pretty much the first thing I tried to test out compilation has some unexpected behavior. I didn't want to report it as a bug, as I don't know the tool at all, but it _seems_ broken...

* `npx create-turbo@latest`
* choose `npm` when prompted for the package manager
* `npm run build` - completes successfully
* edit `apps/web/pages/index.tsx` and add a broken type (by default, there is no explicit type), e.g.
```diff
- export default function Web() {
+ export default function Web(): number {
```
* `npm run build` - this fails as I would expect:
```
web:build: Type error: Type 'Element' is not assignable to type 'number'.
web:build: 
web:build:   2 | 
web:build:   3 | export default function Web(): number {
web:build: > 4 |   return (
web:build:     |   ^
web:build:   5 |     <div>
web:build:   6 |       <h1>Web</h1>
web:build:   7 |       <Button />
```
* revert the file back to original
```diff
- export default function Web(): number {
+ export default function Web() {
```
* `npm run build` - continues to report the type error above
```
web:build: Type error: Type 'Element' is not assignable to type 'number'.
web:build: 
web:build:   2 | 
web:build:   3 | export default function Web(): number {
web:build: > 4 |   return (
web:build:     |   ^web:build:   5 |     <div>
web:build:   6 |       <h1>Web</h1>
web:build:   7 |       <Button />

 Tasks:    2 successful, 2 total
Cached:    2 cached, 2 total
  Time:    847ms >>> FULL TURBO
```
* note that while it reports a type error, the final output indicates success. This is technically correct, as the code on disk now **should** compile. It also returns a zero exit code:
```shell
$ echo $?
0
```
* add a **different** broken type:
```diff
- export default function Web() {
+ export default function Web(): boolean {
```
* `npm run build` - fails as expected:
```
web:build: Type error: Type 'Element' is not assignable to type 'boolean'.
web:build: 
web:build:   2 | 
web:build:   3 | export default function Web(): boolean {
web:build: > 4 |   return (
web:build:     |   ^
```
* revert the file back to original (no explicit type)
```diff
- export default function Web(): boolean {
+ export default function Web() {
```
* `npm run build` - reports a type error, but surprisingly reverts back to reporting about `number`
```
web:build: Type error: Type 'Element' is not assignable to type 'number'.
web:build: 
web:build:   2 | 
web:build:   3 | export default function Web(): number {
web:build: > 4 |   return (
web:build:     |   ^
web:build:   5 |     <div>
web:build:   6 |       <h1>Web</h1>
web:build:   7 |       <Button />

 Tasks:    2 successful, 2 total
Cached:    2 cached, 2 total
  Time:    749ms >>> FULL TURBO
```
---
The last part in particular was surprising. It clearly picked up the change when I used `boolean`, but upon removing it, it seems as though the version with `number` was still cached... somewhere. Other than the output, it seems to be working correctly. Is the actual **output** cached somewhere, perhaps? Grepping around in the project shows numerous binary files that match `function Web()`, but nothing that matches `function Web(): number`.

The version of the CLI is 1.1.5. 
```shell
npm ls turbo
turborepo-basic-shared@0.0.0 /Users/ehaynes/workspace/tmp/turborepo-test
└── turbo@1.1.5
```

**Comment by anthonyshew** [comment, marked as answer]
This may have been a hashing bug.

Marking as outdated seeing that the version is 1.1.5.
